### PR TITLE
Subaru: move to flags within PlatformConfig

### PR DESCRIPTION
**Pull request by jnewb1**
any thoughts on this?

### CODE REVIEW

File: selfdrive/car/subaru/values.py [hunk 1]
```diff
@@ -53,8 +53,13 @@ def __init__(self, CP):
 
 
 class SubaruFlags(IntFlag):
-  SEND_INFOTAINMENT = 1
```
Comment (adeebshihadeh) [inline selfdrive/car/subaru/values.py:56]: it's useful to keep these static so that we can easily determine the flags from the logs

**Review by adeebshihadeh — COMMENTED** [review]
I like this, but it's not all there yet. For example, it's not obvious that GEN1 and GEN2 are mutually exclusive. To make this super clear, we probably want to require specifying the ADAS platform explicitly for every PlatformConfig, e.g. through a subclass.

**Comment by jnewb1** [comment]
honda will need this framework first, because it uses the HONDA_BOSCH set within CarInfo

File: selfdrive/car/subaru/values.py [hunk 2]
```diff
@@ -128,76 +135,87 @@ class CAR(Platforms):
...
 
 
-LKAS_ANGLE = {CAR.FORESTER_2022, CAR.OUTBACK_2023, CAR.ASCENT_2023}
-GLOBAL_GEN2 = {CAR.OUTBACK, CAR.LEGACY, CAR.OUTBACK_2023, CAR.ASCENT_2023}
-PREGLOBAL_CARS = {CAR.FORESTER_PREGLOBAL, CAR.LEGACY_PREGLOBAL, CAR.OUTBACK_PREGLOBAL, CAR.OUTBACK_PREGLOBAL_2018}
-HYBRID_CARS = {CAR.CROSSTREK_HYBRID, CAR.FORESTER_HYBRID}
+LKAS_ANGLE = CAR.with_flags(SubaruFlags.LKAS_ANGLE)
+GLOBAL_GEN2 = CAR.with_flags(SubaruFlags.GLOBAL_GEN2)
+PREGLOBAL_CARS = CAR.with_flags(SubaruFlags.PREGLOBAL)
+HYBRID_CARS = CAR.with_flags(SubaruFlags.HYBRID)
```
Comment (sshane) [inline selfdrive/car/subaru/values.py:214]: Any reason to keep this around if these flags will be statically set (needed for docs with some brands)?

Comment (jnewb1) [reply selfdrive/car/subaru/values.py:214]: was just trying to keep the diff small but we shouldn't need these sets anymore

Comment (jnewb1) [reply selfdrive/car/subaru/values.py]: removed the sets

**Review by jnewb1 — COMMENTED** [review]
```python
class Freezable:
```

we can maintain the freezable nature while still allowing for \_\_post_init\_\_

```
Traceback (most recent call last):
  File "/home/batman/openpilot/selfdrive/car/chrysler/values.py", line 101, in <module>
    CAR.RAM_HD.config.platform_str = 5
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/batman/openpilot/openpilot/selfdrive/car/__init__.py", line 260, in __setattr__
    raise Exception("cannot modify frozen object")
Exception: cannot modify frozen object
```

File: selfdrive/car/subaru/values.py [hunk 3]
```diff
@@ -89,10 +97,14 @@ def init_make(self, CP: car.CarParams):
       self.footnotes.append(Footnote.EXP_LONG)
 
 
-@dataclass(frozen=True)
+@dataclass
 class SubaruPlatformConfig(PlatformConfig):
   dbc_dict: DbcDict = field(default_factory=lambda: dbc_dict('subaru_global_2017_generated', None))
 
+  def init(self):
```
Comment (jnewb1) [inline selfdrive/car/subaru/values.py:104]: @sshane this should help with some of the toyota copy/paste, we can set the DBC based on a flag ex: nodsu

File: common/utils.py [hunk 4]
```diff
@@ -0,0 +1,11 @@
+class Freezable:
+  FROZEN: bool = False
```
Comment (sshane) [inline common/utils.py:2]:
```suggestion
  _frozen: bool = False
```

seems better?

File: selfdrive/car/__init__.py [hunk 5]
```diff
@@ -256,20 +258,30 @@ class CarSpecs:
...
+@dataclass(order=True)
+class PlatformConfig(Freezable):
   platform_str: str
   car_info: CarInfos
   dbc_dict: DbcDict
+  flags: int = 0
 
   specs: CarSpecs | None = None
 
+  FROZEN: bool = False
```
Comment (sshane) [inline selfdrive/car/__init__.py:270]: do you have to specify it here too?

File: selfdrive/car/subaru/values.py (hunk 3 above)
Comment (sshane) [reply selfdrive/car/subaru/values.py:104]: Nice, if there's going to be less grouping, I wanted a way to easily show exceptions like this where one or two platforms may differ from the rest.

File: selfdrive/car/__init__.py (hunk 5 above)
Comment (jnewb1) [reply selfdrive/car/__init__.py]: removed

File: common/utils.py (hunk 4 above)
Comment (jnewb1) [reply common/utils.py]: yea, done.

**Comment by jnewb1** [comment]
going to merge this and try with toyota, if we don't like it we can revert or change stuff

**Comment by sshane** [comment]
I can do the Toyota one with these changes and see how it goes

File: selfdrive/car/subaru/values.py [hunk 6]
```diff
@@ -105,11 +117,13 @@ class CAR(Platforms):
     "SUBARU OUTBACK 6TH GEN",
     SubaruCarInfo("Subaru Outback 2020-22", "All", car_parts=CarParts.common([CarHarness.subaru_b])),
     specs=CarSpecs(mass=1568, wheelbase=2.67, steerRatio=17),
+    flags=SubaruFlags.GLOBAL_GEN2 | SubaruFlags.STEER_RATE_LIMITED,
```
Comment (sshane) [inline selfdrive/car/subaru/values.py:120]: Should we add an `adas_platform` field that is an enum not a flag? You can set GEN2 and PREGLOBAL in one car, and we have no tests to catch this

File: selfdrive/car/subaru/values.py [hunk 7]
```diff
@@ -128,77 +142,77 @@ class CAR(Platforms):
...
 
 
-LKAS_ANGLE = {CAR.FORESTER_2022, CAR.OUTBACK_2023, CAR.ASCENT_2023}
-GLOBAL_GEN2 = {CAR.OUTBACK, CAR.LEGACY, CAR.OUTBACK_2023, CAR.ASCENT_2023}
-PREGLOBAL_CARS = {CAR.FORESTER_PREGLOBAL, CAR.LEGACY_PREGLOBAL, CAR.OUTBACK_PREGLOBAL, CAR.OUTBACK_PREGLOBAL_2018}
-HYBRID_CARS = {CAR.CROSSTREK_HYBRID, CAR.FORESTER_HYBRID}
-
-# Cars that temporarily fault when steering angle rate is greater than some threshold.
-# Appears to be all torque-based cars produced around 2019 - present
-STEER_RATE_LIMITED = GLOBAL_GEN2 | {CAR.IMPREZA_2020, CAR.FORESTER}
```
Comment (sshane) [inline selfdrive/car/subaru/values.py:200]: This PR changes the cars in the STEER_RATE_LIMITED set, this is one of the things we can't regress with this refactor.

```
assert STEER_RATE_LIMITED == CAR.with_flags(SubaruFlags.STEER_RATE_LIMITED)
```

File: selfdrive/car/subaru/values.py (hunk 6 above)
Comment (sshane) [reply selfdrive/car/subaru/values.py:120]: We can just sub-class for this

**Pull request merged**